hoomd/array.py

```diff
@@ -0,0 +1,652 @@
+_ndarray_std_funcs_ = ([
+    # Select subset of array
+    'diagonal',
+    # Reshapes array
+    'reshape', 'transpose', 'swapaxes', 'ravel', 'squeeze',
+], _std_func_with_check)
+
+# Functions that we disallow use of
+
+
+def _disallowed_wrap(method):
+    def raise_error(*args, **kwargs):
+        raise HOOMDArrayError(
+            "The {} method is not allowed for {} objects.".format(
+                method, self.__class__))
+
+    return raise_error
+
+
+_ndarray_disallow_funcs_ = ([
+    'view', 'resize', 'flat', 'flatiter'
+], _disallowed_wrap)
+
+# Properties that can return an array pointing to the same buffer
+
+
+def _wrap_properties_with_check(prop, cls=np.ndarray):
+    prop = getattr(cls, prop)
+
+    @property
+    @functools.wraps(prop)
+    def wrapped(self):
+        arr = self._coerce_to_ndarray()
+        rtn = getattr(arr, prop)
+        if np.may_share_memory(rtn, arr):
+            return self.__class__(rtn, self._callback, self.read_only)
+        else:
+            return rtn
+
+    @wrapped.setter
+    def wrapped(self, value):
+        arr = self._coerce_to_ndarray()
+        return setattr(arr, value)
+
+    return wrapped
+
+
+_ndarray_properties_ = ([
+    'T'
+], _wrap_properties_with_check)
+
+# Properties we disallow access of
+
+
+def _disallowed_property_wrap(method):
+
+    @property
+    def raise_error(self):
+        raise HOOMDArrayError(
+            "The {} property is not allowed for {} objects.".format(
+                method, self.__class__))
+
+    return raise_error
+
+
+_ndarray_disallow_properties_ = ([
+    'data', 'base'
+], _disallowed_property_wrap)
+
+
+_wrap_list = [
+    _ndarray_ops_,
+    _ndarray_magic_safe_,
+    _ndarray_magic_unsafe_,
+    _ndarray_iops_,
+    _ndarray_std_funcs_,
+    _ndarray_disallow_funcs_,
+    _ndarray_properties_,
+    _ndarray_disallow_properties_,
+]
+
+
+def coerce_mock_to_array(val):
+    """Helper function for ``__array_{ufunc,function}__``.
+
+    Coerces ``HOOMDArray`` objects into ``numpy.ndarray`` objects.
+    """
+    if isinstance(val, Iterable) and not isinstance(val,
+                                                    (np.ndarray, HOOMDArray)):
+        return [coerce_mock_to_array(v) for v in val]
+    return val if not isinstance(val, HOOMDArray) else val._coerce_to_ndarray()
+
+
+class HOOMDArray(metaclass=_WrapClassFactory(_wrap_list)):
+    """A NumPy like interface to internal HOOMD-blue data.
+
+    These objects are returned by HOOMD-blue's zero copy access to system data.
+    This class acts like a `numpy.ndarray` object through NumPy's provided
+    `interface <https://numpy.org/doc/stable/reference/arrays.classes.html>`_.
+    Some excepts are the ``view``, ``resize``, ``flat`` and ``flatiter`` methods
+    and the ``data`` and ``base`` properties.  For typical use cases,
+    understanding this class is not necessary. Treat it as a ``numpy.ndarray``.
+
+    We attempt to escape this class whenever possible. This essentially means
```

Comment:
The "escape" terminology is confusing and I would remove it, just start with the following sentence (that you try to make a numpy array ASAP). Also remove the reference to escaping in the next paragraph.